### PR TITLE
Risky spl detector

### PR DESCRIPTION
**Pull request by ksharad-splunk**
# PR Template for new Detections

### Details

risky_spl_detection

### Author Checklist

- [ ] Validate name matches `<platform>_<mitre att&ck technique>_<short description>`
- [ ] Make sure that CI/CD [detection-testing and build-and-validate](https://github.com/splunk/security_content/actions) jobs passed ✔️ 
- [ ] Is there an Atomic Test? Is GUID on the test file under array: `atomic_test_guid`, [example]()?

### Review Checklist

- [ ] Validated SPL logic.
- [ ] Validated tags, description, and how to implement.
- [ ] Verified references match analytic.

### CODE REVIEW

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml [hunk 1]
```diff
@@ -0,0 +1,58 @@
+name: Splunk Command and Scripting Interpreter Detect Risky SPL MLTK
+id: b4aefb5f-1037-410d-a149-1e091288ba33
+version: 1
+date: '2022-06-16'
+author: Abhinav Mishra, Kumar Sharad, Namratha Sreekanta and Xiao Lin, Splunk
+type: Anomaly
+datamodel:
+- Splunk_Audit
+description: 'The following analytic uses a pre-trained machine learning text classifier to detect potentially risky commands. A command is deemed risky based on the presence of certain trigger keywords, along with the context and the role of the user (please see references). The model takes as input the command text, user and search type and outputs a risk score between [0,1]. A high score indicates higher likelihood of a command being risky.'
+search: '| tstats `security_content_summariesonly` count min(_time) as firstTime max(_time) as lastTime from datamodel=Splunk_Audit.Search_Activity by Search_Activity.search Search_Activity.info Search_Activity.total_run_time Search_Activity.user Search_Activity.savedsearch_name Search_Activity.search_type
```
Comment (trogdorsey) [inline detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: There is a lot of stuff in the by clause that isn't used further down in the search. If those fields are not needed, then they should be removed.

File: tests/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.test.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+name: Splunk Command and Scripting Interpreter Detect Risky SPL MLTK Unit Test
+tests:
+- name: Splunk Command and Scripting Interpreter Detect Risky SPL MLTK
+  file: application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml
+  pass_condition: '| stats count | where count > 0'
```
Comment (trogdorsey) [inline tests/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.test.yml]: pass condition should not be > 0.  It's a best practice to set it to the exact number of anomalies that are expected to be found in the test data

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+  | eval spl_text = 'Search_Activity.search'. " " .'Search_Activity.user'. " " .'Search_Activity.search_type'
+  | dedup spl_text
+  | apply DetectRiskySPL
+  | table Search_Activity.search, Search_Activity.user, Search_Activity.search_type, risk_score
```
Comment (trogdorsey) [inline detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: You should drop the data model object before you output the table. Something like `drop_dm_object_name("Search_Activity")`

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+id: b4aefb5f-1037-410d-a149-1e091288ba33
+version: 1
+date: '2022-06-16'
+author: Abhinav Mishra, Kumar Sharad, Namratha Sreekanta and Xiao Lin, Splunk
+type: Anomaly
+datamodel:
+- Splunk_Audit
+description: 'The following analytic uses a pre-trained machine learning text classifier to detect potentially risky commands. A command is deemed risky based on the presence of certain trigger keywords, along with the context and the role of the user (please see references). The model takes as input the command text, user and search type and outputs a risk score between [0,1]. A high score indicates higher likelihood of a command being risky.'
+search: '| tstats `security_content_summariesonly` count min(_time) as firstTime max(_time) as lastTime from datamodel=Splunk_Audit.Search_Activity by Search_Activity.search Search_Activity.info Search_Activity.total_run_time Search_Activity.user Search_Activity.savedsearch_name Search_Activity.search_type
+  | eval spl_text = 'Search_Activity.search'. " " .'Search_Activity.user'. " " .'Search_Activity.search_type'
```
Comment (trogdorsey) [inline detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: I'm guessing the model uses spl_text as an input? Why not just the 3 fields separately?

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 3 above)
Comment (trogdorsey) [inline detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: There is no filtering based on risk score, so a risk event will be created for every event, which isn't what we want here.

**Review by trogdorsey — DISMISSED** [review]

**Comment by patel-bhavin** [comment]
A few thoughts:

- Can you please add a screenshot of the search execution to this PR? 

- Also, the search name is almost like the one we already released. https://github.com/splunk/security_content/blob/develop/detections/application/splunk_command_and_scripting_interpreter_risky_spl_mltk.yml

We probably should update the search name to something more specific to the model being used. 

- This PR needs a baseline yml file that uses fit command

**Comment by trogdorsey** [comment]
> * This PR needs a baseline yml file that uses fit command

No baseline yml is needed. It is shipping a pretrained model.

**Comment by patel-bhavin** [comment]
> > * This PR needs a baseline yml file that uses fit command
> 
> No baseline yml is needed. It is shipping a pretrained model.

oh damn! my bad.  thx

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 3 above)
Comment (ksharad-splunk) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: Done!

Comment (ksharad-splunk) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: Done!

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 4 above)
Comment (ksharad-splunk) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: It doesn't make a difference to the model internally.

**Comment by ksharad-splunk** [comment]
> A few thoughts:
> 
> * Can you please add a screenshot of the search execution to this PR?
> * Also, the search name is almost like the one we already released. https://github.com/splunk/security_content/blob/develop/detections/application/splunk_command_and_scripting_interpreter_risky_spl_mltk.yml
> 
> We probably should update the search name to something more specific to the model being used.

Added the screenshot, we can talk about the name.

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 3 above)
Comment (trogdorsey) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: the drop data model object needs to be before the table command. And then you will need to remove Search_Activity from the field names.

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+search: '| tstats `security_content_summariesonly` count min(_time) as firstTime max(_time) as lastTime from datamodel=Splunk_Audit.Search_Activity where Search_Activity.search_type=adhoc Search_Activity.user!=splunk-system-user by Search_Activity.search Search_Activity.user Search_Activity.search_type
+  | eval spl_text = 'Search_Activity.search'. " " .'Search_Activity.user'. " " .'Search_Activity.search_type'
+  | dedup spl_text
+  | apply DetectRiskySPL
+  | where risk_score > 0.6
+  | table Search_Activity.search, Search_Activity.user, Search_Activity.search_type, risk_score
+  | `drop_dm_object_name(Search_Activity)`
+  |`splunk_command_and_scripting_interpreter_detect_risky_spl_mltk_filter`'
+how_to_implement: This detection depends on the MLTK app which can be found here - https://splunkbase.splunk.com/app/2890/ and the Splunk Audit datamodel which can be found here - https://splunkbase.splunk.com/app/1621/. Additionally, you need to be ingesting logs which include Search_Activity.search, Search_Activity.user, Search_Activity.search_type from your endpoints.
```
Comment (trogdorsey) [inline detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: Minor thing here, but might be worth mentioning changing the risk_score threshold based on their environments.

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 3 above)
Comment (ksharad-splunk) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: Updated.

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 5 above)
Comment (ksharad-splunk) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: Updated.

File: detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml (hunk 1 above)
Comment (patel-bhavin) [reply detections/application/splunk_command_and_scripting_interpreter_detect_risky_spl_mltk.yml]: this has been addressed

File: detections/application/detect_risky_spl_using_pretrained_ml_model.yml [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  cve:
+  - CVE-2022-32154
+  context:
+  - Source:Endpoint
+  dataset:
+  - https://github.com/splunk/attack_data/raw/master/datasets/attack_techniques/T1203/search_activity.txt
+  impact: 50
+  kill_chain_phases:
+  - Actions on Objectives
+  message: A potentially risky Splunk command has been run by $(Search_Activity.user), kindly review.
```
Comment (trogdorsey) [inline detections/application/detect_risky_spl_using_pretrained_ml_model.yml]: This should just be user

Comment (ksharad-splunk) [reply detections/application/detect_risky_spl_using_pretrained_ml_model.yml]: Done!

**Review by josehelps — CHANGES_REQUESTED** [review]
please specific in the description this is a on-prem only model.

**Review by patel-bhavin — APPROVED** [review]

**Review by patel-bhavin — APPROVED** [review]

**Review by josehelps — APPROVED** [review]
LGTM :smile: awesome work @ksharad-splunk

**Pull request merged**